api/src/main/java/org/infinispan/api/async/Cache.java

```diff
@@ -0,0 +1,167 @@
+package org.infinispan.api.async;
+
+import java.util.EnumSet;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.CompletionStage;
+import java.util.concurrent.Flow;
+import java.util.function.BiFunction;
+import java.util.function.Function;
+
+import org.infinispan.api.common.tasks.EntryConsumerTask;
+import org.infinispan.api.common.Flag;
+import org.infinispan.api.common.events.AsyncListenerHandle;
+import org.infinispan.api.common.events.KeyValueListener;
+import org.infinispan.api.configuration.CacheConfiguration;
+
+/**
+ * @since 13.0
+ **/
+public interface Cache<K, V> {
+   /**
+    * The name of the cache.
+    *
+    * @return
+    */
+   String name();
+
+   /**
+    * The configuration for this cache.
+    *
+    * @return
+    */
+   CompletionStage<CacheConfiguration> configuration();
+
+   /**
+    * Returns a new instance of this cache with the supplied {@link Flag}s
+    * @param flags
+    * @return
+    */
+   Cache<K, V> withFlags(EnumSet<Flag> flags);
+
+   /**
+    * Returns a new instance of this cache with the supplied {@link Flag}s
+    * @param flags
+    * @return
+    */
+   Cache<K, V> withFlags(Flag... flags);
+
+   /**
+    * Get the value of the Key if such exists
+    *
+    * @param key
+    * @return the value
+    */
+   CompletionStage<V> get(K key);
+
+   /**
+    * Insert the key/value if such key does not exist
+    *
+    * @param key
+    * @param value
+    * @return Void
+    */
+   CompletionStage<Boolean> putIfAbsent(K key, V value);
+
+   /**
+    * @param key
+    * @param value
+    * @return Void
+    */
+   CompletionStage<V> put(K key, V value);
+
+   /**
+    * @param key
+    * @param value
+    * @return
+    */
+   CompletionStage<Void> set(K key, V value);
+
+   /**
+    * Delete the key
+    *
+    * @param key
+    * @return whether the entry was removed.
+    */
+   CompletionStage<Boolean> remove(K key);
```

Comment:
Will we also have a remove variant that returns CompletionStage<V>?  Or would I be expected to use a compute(...) whose function returns null?

Reply:
I've added a `getAndRemove`. The name isn't great.